linkerd/router/src/cache.rs

```diff
@@ -0,0 +1,89 @@
+use crate::NewService;
+use ahash::AHashMap;
+use parking_lot::Mutex;
+use std::{fmt::Debug, hash::Hash, marker::PhantomData, sync::Arc};
+
+/// A [`NewService`] that lazy builds stacks for each `K`-typed key.
+#[derive(Debug)]
+pub struct NewCache<K, N> {
+    inner: N,
+    _marker: PhantomData<fn(K)>,
+}
+
+/// A [`NewService`] that lazily builds an inner `S`-typed service for each
+/// `K`-typed target.
+#[derive(Debug)]
+pub struct Cache<K, N, S> {
+    new_inner: N,
+    services: Arc<Mutex<AHashMap<K, S>>>,
+}
+
+// === impl NewCache ===
+
+impl<K, N> NewCache<K, N> {
+    pub fn new(inner: N) -> Self {
+        Self {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T, K, N, M> NewService<T> for NewCache<K, N>
+where
+    T: Clone,
+    N: NewService<T, Service = M>,
+    M: NewService<K>,
+{
+    type Service = Cache<K, M, M::Service>;
+
+    fn new_service(&self, target: T) -> Self::Service {
```

Comment:
nit: should maybe inline?
```suggestion
    #[inline]
    fn new_service(&self, target: T) -> Self::Service {
```